### PR TITLE
OpenVPN Support with GitHub OAuth2 +Duo MFA Authorization

### PR DESCRIPTION
**Pull request by goruha**
## what
* Implement OpenVPN with Github OAuth Authorization and Duo for MFA (via pam)

## why
* Secure remote cluster access
* Centralized user management
* Multi-factor Authentication by Github & Duo requires that both a MFA device and enforces a sort of [Two-Man Rule](https://en.wikipedia.org/wiki/Two-man_rule) to grant access
* Duo provides audit logs and IP geo restrictions

## todo
- [x] add ENV for github teams (CSV)
- [x] finish openvpn-api integration
- [x] add pam duo support
- [x] make cert expirations configurable; it's probably one of these settings for the `openvpn-api` container:
   ```
   set_var EASYRSA_CA_EXPIRE	3650
   set_var EASYRSA_CERT_EXPIRE    3650
   set_var EASYRSA_CRL_DAYS    180
   ```
- [x] address PR review: https://github.com/cloudposse/openvpn-api/pull/1
- [x] reduce re-authentication interval (make it configurable via `values.yaml`); 
   ```
   # disable username/password renegotiation
   reneg-sec 0
   ```
- [x] pin images to `latest`

## who
@cloudposse/engineering

### CODE REVIEW

File: incubator/bastion/templates/deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+  template:
+    metadata:
+      labels:
+        app: {{ template "fullname" . }}
+      annotations:
+        pod.beta.kubernetes.io/init-containers: '[{
+          "name": "init",
+          "image": "{{ .Values.image.repository }}:{{ .Values.image.tag }}",
+          "imagePullPolicy": {{ .Values.image.pullPolicy | quote }},
+          "command": ["/bin/sh", "-c", "(cp -a /etc/ /mnt/etc/) && (mkdir -p /mnt/home /mnt/root /mnt/var/log)"],
```
Comment (osterman) [inline incubator/bastion/templates/deployment.yaml]: I think you can drop `mkdir /mnt/home /mnt/root /mnt/var/log`

Comment (osterman) [inline incubator/bastion/templates/deployment.yaml]: oh wait... why are we modifying `incubator/bastion/templates/deployment.yaml` in this PR? =)

**Comment by osterman** [comment]
We should also expose these:
```
        set_var EASYRSA_REQ_COUNTRY	"US"
	set_var EASYRSA_REQ_PROVINCE	"California"
	set_var EASYRSA_REQ_CITY	"San Francisco"
	set_var EASYRSA_REQ_ORG		"Copyleft Certificate Co"
	set_var EASYRSA_REQ_EMAIL	me@example.net
	set_var EASYRSA_REQ_OU		"My Organizational Unit"
```

**Comment by osterman** [comment]
Actually, cannot define using ENVs. Need to use these args I think:

https://github.com/OpenVPN/easy-rsa/blob/master/easyrsa3/easyrsa#L1086

`set_var` overwrites the existing envs. :(
image

File: incubator/openvpn/templates/vpn.deployment.yaml [hunk 2]
```diff
@@ -20,13 +20,18 @@ spec:
                   "env": [
                     { "name": "OVPN_OTP_AUTH", "value": "1" },
                     { "name": "EASYRSA_BATCH", "value": "1" },
+                    { "name": "EASYRSA_VARS_FILE", "value": "/etc/easyrsa.d/vars" },
```
Comment (osterman) [inline incubator/openvpn/templates/vpn.deployment.yaml]: nice!

File: incubator/openvpn/templates/vpn.configmap.yaml [hunk 3]
```diff
@@ -1,4 +1,29 @@
...
+kind: ConfigMap
+metadata:
+  name: "{{ template "fullname" . }}"
+  labels:
+    app: {{ template "fullname" . }}
+    chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
+    release: "{{ .Release.Name }}"
+    heritage: "{{ .Release.Service }}"
+data:
+  vars: |-
```
Comment (osterman) [inline incubator/openvpn/templates/vpn.configmap.yaml]: nice!

File: incubator/openvpn/templates/vpn.service.yaml [hunk 4]
```diff
@@ -15,7 +15,7 @@ spec:
     targetPort: 1194
     protocol: TCP
     name: "{{ .Values.openvpn.service.name }}-vpn"
-{{ if ( and .Values.ui.enabled .Values.ui.ssl.enabled ) }}
+{{ if .Values.ui.ssl.enabled }}
```
Comment (osterman) [inline incubator/openvpn/templates/vpn.service.yaml]: I don't understand what this has to do with `ssl.enabled`.

Comment (osterman) [inline incubator/openvpn/templates/vpn.service.yaml]: Should probably just be dropped from here

File: incubator/openvpn/values.yaml [hunk 5]
```diff
@@ -47,9 +46,22 @@ persistence:
...
+  vars:
+    EASYRSA_REQ_COUNTRY: "US"
+    EASYRSA_REQ_PROVINCE: "California"
+    EASYRSA_REQ_CITY: "San Francisco"
+    EASYRSA_REQ_ORG: "Copyleft Certificate Co"
+    EASYRSA_REQ_EMAIL: "me@example.net"
+    EASYRSA_REQ_OU: "My Organizational Unit"
+    EASYRSA_CA_EXPIRE: "3650"
+    EASYRSA_CERT_EXPIRE: "3650"
+    EASYRSA_CRL_DAYS: "180"
```
Comment (osterman) [inline incubator/openvpn/values.yaml]: I think we should just move `OVPN_RENEG_SEC` here. We can default it to 0. No need to maintain it separately. You can also remove it from the "special" section at the top of `values.yaml`

File: incubator/openvpn/values.yaml [hunk 6]
```diff
@@ -47,9 +46,22 @@ persistence:
...
+
 debug: 0
 
 openvpn:
+  renegotiateConnectionAfter: *RENEGOTIATE_CONNECTION_AFTER
```
Comment (osterman) [inline incubator/openvpn/values.yaml]: Should we just move OVPN_RENEG_SEC here in an `env` section? We can default it to 0. No need to maintain it separately. You can also remove it from the "special" section at the top of values.yaml

File: incubator/openvpn/templates/vpn.service.yaml (hunk 4 above)
Comment (goruha) [inline incubator/openvpn/templates/vpn.service.yaml]: If ssl enabled service should expose 80 port (required for let's encrypt)

Comment (osterman) [inline incubator/openvpn/templates/vpn.service.yaml]: shouldn't we always expose it even if ssl disabled?

Comment (goruha) [inline incubator/openvpn/templates/vpn.service.yaml]: no, we should not.

File: incubator/openvpn/values.yaml (hunk 6 above)
Comment (goruha) [inline incubator/openvpn/values.yaml]: Done

**Pull request merged**